nlmod/dims/time.py

```diff
@@ -127,6 +136,99 @@ def set_ds_time(
     return ds
 
 
+def set_ds_time(ds, time, start, steady=True, time_units="DAYS", nstp=1, tsmult=1.0):
+    """Set time discretisation for model dataset.
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        model dataset
+    time : array-like
+        array-like of floats (indicating elapsed time) or timestamps corresponding to
+        the end of each stress period in the model.
+    start : str or pandas.Timestamp, optional
+        model start datetime as string or pandas Timestamp, if None, defaults to
+        1 january 2000.
+    steady : arraylike or bool, optional
+        arraylike indicating which stress periods are steady-state, by default True,
+        which sets all stress periods to steady-state.
+    time_units : str, optional
+        time units, by default "DAYS"
+    nstp : int or array-like, optional
+        number of steps per stress period, stored in ds.attrs, default is 1
+    tsmult : float, optional
+        timestep multiplier within stress periods, stored in ds.attrs, default is 1.0
+
+    Returns
+    -------
+    ds : xarray.Dataset
+        model dataset with added time coordinate
+
+    """
+    logger.info(
+        "This is the new version of set_ds_time()."
+        " If you're looking for the old behavior,"
+        "use `nlmod.time.set_ds_time_deprecated()`."
+    )
+
+    # parse start datetime
+    if isinstance(start, str):
+        start = pd.Timestamp(start)
+    elif isinstance(start, (pd.Timestamp, np.datetime64)):
+        pass
+    else:
+        raise TypeError("Cannot parse start datetime.")
+
+    # convert time to Timestamps
+    # calculate time idx
+    if isinstance(time[0], (int, np.integer, float)):
+        time = pd.Timestamp(start) + pd.to_timedelta(time, time_units)
+    elif isinstance(time[0], str):
+        time = pd.to_datetime(time)
+    elif isinstance(time[0], (pd.Timestamp, np.datetime64, xr.core.variable.Variable)):
+        pass
+    else:
+        raise TypeError("Cannot process 'time' argument. Datatype not understood.")
+
+    # set steady
+    if isinstance(steady, bool):
+        steady = steady * np.ones(len(time))
+
+    ds = ds.assign_coords(coords={"time": time})
+    if time_units == "D":
+        time_units = "DAYS"
+    ds.time.attrs["time_units"] = time_units
+    ds.time.attrs["start"] = str(start)
+    ds.time.attrs["steady"] = steady
```

Comment:
I think steady is always a boolean numpy array. We could also add this as a DataVariable to the Dataset with dimension time.

Reply:
Good idea, I like that.